File: algorithms/selected_sort.js
var swap = function(array, firstIndex, secondIndex) {
    var temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
};

var indexOfMinimum = function(array, startIndex) {

    var minValue = array[startIndex];
    var minIndex = startIndex;

    for(var i = minIndex + 1; i < array.length; i++) {
        if(array[i] < minValue) {
            minIndex = i;
            minValue = array[i];
        }
    }
    return minIndex;
};

var selectionSort = function(array) {
    var indexValue;
    for(var i = 0; i < array.length; i ++){
        indexValue = indexOfMinimum(array, i);
        swap(array, indexValue, i);
    }
};

var array = [22, 11, 99, 88, 9, 7, 42];
selectionSort(array);
println("Array after sorting:  " + array);

Program.assertEqual(array, [7, 9, 11, 22, 42, 88, 99]);
var secondArray = [8, 12, 4, 2, 22, 9, 18, 99, 21];
selectionSort(secondArray);
Program.assertEqual(secondArray, [2, 4, 8, 9, 12, 18, 21, 22, 99]);

File: algorithms/selected_sort.rb
def index_of_minimum(array, start_index)
  min_value = array[start_index]
  min_index = start_index
  index_counter = min_index + 1
  while index_counter < array.length do
    if array[index_counter] < min_value
      min_index = index_counter
      min_value = array[index_counter]
    end
    index_counter += 1
  end
  min_index
end

def swap (array, first_index, second_index)
  temp = array[first_index]
  array[first_index] = array[second_index]
  array[second_index] = temp
end

def selected_sort(array)
  index_value = 0
  i = 0
  while i < array.length do
    index_value = index_of_minimum(array, i)
    swap(array, index_value, i)
    i += 1
  end
  p array
end

array = [22, 11, 99, 88, 9, 7, 42]
selected_sort(array)
